lib/resty/limit/conn.lua

```diff
@@ -47,58 +47,41 @@ function _M.incoming(self, key, commit)
     local max = self.max
 
     self.committed = false
+    if commit then
+        local conn, err = dict:incr(key, 1, 0)
+        if not conn then
+            return nil, err
+        end
 
-    local conn = dict:get(key)
-    if conn then
-        if conn >= max then
-            if conn >= max + self.burst then
-                return nil, "rejected"
-            end
-
-            if commit then
-                local err
-                conn, err = dict:incr(key, 1, 0)
-                if not conn then
-                    return nil, err
-                end
-
-                self.committed = true
-            else
-                conn = conn + 1
+        if conn > max + self.burst then
+            conn, err = dict:incr(key, -1, conn)
+            if not conn then
+                return nil, err
             end
+            return nil, "rejected"
+        end
+        self.committed = true
 
+        if conn > max then
             -- make the exessive connections wait
             return self.unit_delay * floor((conn - 1) / max), conn
         end
 
-        if commit then
-            -- FIXME: we should use incr_or_init here.
-            local new_conn, err = dict:incr(key, 1)
-            if not new_conn then
-                return nil, err
-            end
+        -- we return a 0 delay by default
+        return 0, conn
+    end
 
-            self.committed = true
-            return 0, new_conn
+    local conn = dict:get(key) or 0
+    if conn >= max then
+        if conn >= max + self.burst then
```

Comment:
I was about to say, I think we should also nest this case under the `conn > max` branch up there for the `commited` branch, and then I see it is being done here for the non-commited. Should we? It used to be.